File: regions/system/library/csharp/retry/CircuitBreaker.cs
namespace K1s0.System.Retry;

public enum CircuitBreakerState
{
    Closed,
    Open,
    HalfOpen,
}

public class CircuitBreakerConfig
{
    public int FailureThreshold { get; init; } = 5;

    public int SuccessThreshold { get; init; } = 2;

    public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(30);
}

public class CircuitBreaker
{
    private readonly CircuitBreakerConfig _config;
    private readonly SemaphoreSlim _lock = new(1, 1);
    private CircuitBreakerState _state = CircuitBreakerState.Closed;
    private int _failureCount;
    private int _successCount;
    private DateTime _openedAt;

    public CircuitBreaker(CircuitBreakerConfig config)
    {
        _config = config;
    }

    public CircuitBreakerState State => _state;

    public bool IsOpen()
    {
        if (_state == CircuitBreakerState.Open)
        {
            if (DateTime.UtcNow - _openedAt >= _config.Timeout)
            {
                _state = CircuitBreakerState.HalfOpen;
                _successCount = 0;
                return false;
            }

            return true;
        }

        return false;
    }

    public async Task RecordSuccessAsync()
    {
        await _lock.WaitAsync().ConfigureAwait(false);
        try
        {
            if (_state == CircuitBreakerState.HalfOpen)
            {
                _successCount++;
                if (_successCount >= _config.SuccessThreshold)
                {
                    _state = CircuitBreakerState.Closed;
                    _failureCount = 0;
                    _successCount = 0;
                }
            }
            else
            {
                _failureCount = 0;
            }
        }
        finally
        {
            _lock.Release();
        }
    }

    public async Task RecordFailureAsync()
    {
        await _lock.WaitAsync().ConfigureAwait(false);
        try
        {
            if (_state == CircuitBreakerState.HalfOpen)
            {
                _state = CircuitBreakerState.Open;
                _openedAt = DateTime.UtcNow;
                _successCount = 0;
                return;
            }

            _failureCount++;
            if (_failureCount >= _config.FailureThreshold)
            {
                _state = CircuitBreakerState.Open;
                _openedAt = DateTime.UtcNow;
            }
        }
        finally
        {
            _lock.Release();
        }
    }
}

File: regions/system/library/rust/retry/src/circuit_breaker.rs
pub use k1s0_circuit_breaker::CircuitBreakerConfig;
pub use k1s0_circuit_breaker::CircuitBreakerState;

/// Compatibility wrapper that delegates implementation to `k1s0-circuit-breaker`.
pub struct CircuitBreaker {
    inner: k1s0_circuit_breaker::CircuitBreaker,
}

impl CircuitBreaker {
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            inner: k1s0_circuit_breaker::CircuitBreaker::new(config),
        }
    }

    pub async fn is_open(&self) -> bool {
        matches!(self.inner.state().await, CircuitBreakerState::Open)
    }

    pub async fn record_success(&self) {
        self.inner.record_success().await;
    }

    pub async fn record_failure(&self) {
        self.inner.record_failure().await;
    }

    pub async fn get_state(&self) -> CircuitBreakerState {
        self.inner.state().await
    }
}
